pkg/model/multiversion/delta.go

```diff
@@ -0,0 +1,248 @@
+type FieldChangeType string
+
+const (
+	// FieldChangeTypeUnknown is used when ChangeType cannot be computed.
+	FieldChangeTypeUnknown FieldChangeType = "unknown"
+	// FieldChangeTypeIntact is used a field name and structure didn't change.
+	FieldChangeTypeIntact FieldChangeType = "intact"
+	// FieldChangeTypeAdded is used when a new field is introduced in a CRD.
+	FieldChangeTypeAdded FieldChangeType = "added"
+	// FieldChangeTypeRemoved is used a when a field is removed from a CRD.
+	FieldChangeTypeRemoved FieldChangeType = "removed"
+	// FieldChangeTypeRenamed is used when a field is renamed.
+	FieldChangeTypeRenamed FieldChangeType = "renamed"
+	// FieldChangeTypeShapeChanged is used when a field shape has changed.
+	FieldChangeTypeShapeChanged FieldChangeType = "shape-changed"
+	// FieldChangeTypeShapeChangedToSecret is used when a field change to
+	// a k8s secret type.
+	FieldChangeTypeShapeChangedToSecret FieldChangeType = "shape-changed-to-secret"
+)
+
+// FieldDelta represent the delta between the same field in two different
+// CRD versions. If a field is removed in the Hub version the Hub value will
+// be nil. If a field is new in the Hub version, the Spoke value will be nil.
+type FieldDelta struct {
+	ChangeType FieldChangeType
+	// Field from the hub version CRD
+	Hub *ackmodel.Field
+	// Field from the spoke version CRD
+	Spoke *ackmodel.Field
+}
+
+// CRDDelta stores the spec and status deltas for a custom resource.
+type CRDDelta struct {
+	SpecDeltas   []FieldDelta
+	StatusDeltas []FieldDelta
+}
+
+// ComputeCRDFieldsDeltas compares two ackmodel.CRD instances and returns the
+// spec and status fields deltas. spokeCRD is the CRD of the spoke (source) version
+// and hubCRD is the CRD of the hub (destination) version.
+func ComputeCRDFieldsDeltas(spokeCRD, hubCRD *ackmodel.CRD) (*CRDDelta, error) {
+	renames, _, err := hubCRD.GetAllRenames()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get resource field renames: %s", err)
+	}
+
+	specDeltas, err := ComputeFieldsDiff(spokeCRD.SpecFields, hubCRD.SpecFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute spec fields deltas: %s", err)
+	}
+
+	statusDeltas, err := ComputeFieldsDiff(spokeCRD.StatusFields, hubCRD.StatusFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute status fields deltas: %s", err)
+	}
+
+	return &CRDDelta{
+		specDeltas,
+		statusDeltas,
+	}, nil
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// ComputeFieldsDiff computes the difference between two maps of fields. It returns a list
+// of FieldDelta's that contains the the ChangeType and at least one field reference.
+func ComputeFieldsDiff(
+	spokeCRDFields map[string]*ackmodel.Field,
+	hubCRDFields map[string]*ackmodel.Field,
+	// the hub renames
+	renames map[string]string,
+) ([]FieldDelta, error) {
+	// the smallest size of deltas we can get is max(len(spokeCRDFields), len(hubCRDFields))
```

Comment:
do you mean maximum size of delta, instead of smallest?

Reply:
I mean the minimum: if `[A, B]` changes to `[A, B, C]` the delta is `[A->A, B->B, ->C]`

Reply:
I'm removed this size since it's not needed at all